flyrs/src/lib.rs

```diff
@@ -0,0 +1,210 @@
+use tonic::{
+    transport::{Channel, Uri},
+    Request,
+    Response
+};
+
+// Foreign Rust error types: https://pyo3.rs/main/function/error-handling#foreign-rust-error-types
+// Create a newtype wrapper, e.g. MyOtherError. Then implement From<MyOtherError> for PyErr (or PyErrArguments), as well as From<OtherError> for MyOtherError.
+
+// Failed at encoding responded object to bytes string
+struct MessageEncodeError(EncodeError);
+// Failed at decoding requested object from bytes string
+struct MessageDecodeError(DecodeError);
+
+impl fmt::Display for MessageEncodeError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "")
+    }
+}
+
+impl fmt::Display for MessageDecodeError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "")
+    }
+}
+
+impl std::convert::From<MessageEncodeError> for PyErr {
+    fn from(err: MessageEncodeError) -> PyErr {
+        PyOSError::new_err(err.to_string())
+    }
+}
+
+impl std::convert::From<MessageDecodeError> for PyErr {
+    fn from(err: MessageDecodeError) -> PyErr {
+        PyOSError::new_err(err.to_string())
+    }
+}
+
+impl std::convert::From<EncodeError> for MessageEncodeError {
+    fn from(other: EncodeError) -> Self {
+        Self(other)
+    }
+}
+
+impl std::convert::From<DecodeError> for MessageDecodeError {
+    fn from(other: DecodeError) -> Self {
+        Self(other)
+    }
+}
+
+/// A Python class constructs the gRPC service stubs and a Tokio asynchronous runtime in Rust.
+#[pyclass(subclass)]
+pub struct FlyteClient {
+    admin_service: AdminServiceClient<Channel>,
+    runtime: Runtime,
+}
+
+pub fn decode_proto<T>(bytes_obj: &PyBytes) -> Result<T, MessageDecodeError>
+    where
+    T: Message + Default, {
+    let bytes = bytes_obj.as_bytes();
+    let de = Message::decode(&bytes.to_vec()[..]);
+    Ok(de?)
+}
+
+pub fn encode_proto<T>(res:T) -> Result<Vec<u8>, MessageEncodeError>
+where
+T: Message + Default, {
+    let mut buf = vec![];
+    res.encode(&mut buf)?;
+    Ok(buf)
+}
+
+#[pymethods]
+impl FlyteClient {
+    #[new] // Without this, you cannot construct the underlying class in Python.
+    #[pyo3(signature = (endpoint, **kwargs))]
+    pub fn new(endpoint: &str, kwargs: Option<&Bound<'_, PyDict>>) -> PyResult<FlyteClient> {
+        // Use Atomic Reference Counting abstractions as a cheap way to pass string reference into another thread that outlives the scope.
+        let s = Arc::new(endpoint);
+        // Check details for constructing Tokio asynchronous `runtime`: https://docs.rs/tokio/latest/tokio/runtime/struct.Builder.html#method.new_current_thread
+        let rt = match Builder::new_current_thread().enable_all().build() {
+            Ok(rt) => rt,
+            Err(error) => panic!("Failed to initiate Tokio multi-thread runtime: {:?}", error),
+        };
+        // Check details for constructing `channel`: https://docs.rs/tonic/latest/tonic/transport/struct.Channel.html#method.builder
```

Comment:
Currently, a third party library can extend `FlyteRemote` by subclassing it and use `RawSynchronousFlyteClient._channel` directly as the authenticated channel for GRPC calls.

Assuming, we get intercepters + authentication to work in Rust, is there a way to reuse the authenticated channel in a third party library?

For example, can we wrap the channel object and expose it in Python? Then the channel can be passed to other clients that want to use the channel for other services. Concretely:

```python
channel = flyrs.get_authenticated_channel(endpoint)

flytekit_client = flyrs.FlyteClient(channel)
another_client = another_lib.Client(channel)
```

Reply:
It might be worth considering that if the aim is to create a subclass of `FlyteRemote`, it's preferable for those classes to rely on the Flyte client configured in the parent class, rather than assuming that the flyte client is implemented via grpc and assume that there's an underlying `_channel`. Consider for a second the following, we could imagine a scenario where there's a different FlyteClient that doesn't use grpc at all but instead interacts with the HTTP endpoints directly. (Just to clarify, this isn't a reality yet; I'm simply using it as an example to emphasize the importance of exposing a channel as part of the Flyte client.) In that case it's not ok to assume that there's an underlying channel.

You might be talking about the case where someone is implementing a separate flyte client. If that's the case, I think it's fair to assume that they can (should?) bypass our implementation and re-implement auth+grpc.

wdyt?

Reply:
> In that case it's not ok to assume that there's an underlying channel.

This makes sense in terms of the user level `FlyteRemote`. I'm considering the case of extending this PR's `RustFlyteRemote`, which I assume will be GRPC. In my above example, `another_lib.Client` is a external Rust client that uses the same channel as the `RustFlyteRemote`. 

> If that's the case, I think it's fair to assume that they can (should?) bypass our implementation and re-implement auth+grpc.

From a software design point of view, this is reasonable. It matches the current Python implementation which already hides the underlying `_channel`. An external library can create it's own channel, but that will show the authentication screen twice for PCKE auth. (One auth screen for `flytekit` and another one for the external library). To work around this, an external library can reuse the already authenticated `_channel` from `flytekit` and use that for GRPC calls.

---

For this PR, I see three possible paths for external libraries

1. Wrap Rust's GRPC Channel in Python so that external library can use it to make GRPC calls from Rust with the authenticated channel.
2. External library will build it's own `another_lib.RustFlyteClient` that extends `flyrs.RustFlyteClient`. The external library will use `another_lib.RustFlyteClient` instead of `flyters.RustFlyteClient`.
3. Have Auth happen in Python (outside of Rust) and pass in the credentials to `flyrs.RustFlyteClient`. External library will use the same credentials in it's client. This would be nice for HTTP clients as well, because I suspect they will need the same credentials. 

If we can get option 3 to work, I prefer it because it is the most flexible.

Reply:
I like option 3.
We reuse the credentials from the Python (`flytekit`) authenticator and pass them into `flyrs.RustFlyteClient` to wrap it up with Rust grpc channel via Interceptors.

Reply:
I have not tried manually refreshing, but I feel like it will work. Although, if I manually refresh before a GRPC call, I may end up authenticating without needing to make a GRPC call. Currently, we authenticate in a GRPC intercepter to do auth at the very last moment.

For the follow up for this PR and option 3 in https://github.com/flyteorg/flytekit/pull/2377/files#r1589988041, I'll like to do the Auth when `FlyteRemote.client` is created. Concretely, I want to assume that `FlyteRemote.client` is created only if we need to communicate with GRPC.